spanner/value.go

```diff
@@ -877,6 +877,96 @@ func (n *PGNumeric) UnmarshalJSON(payload []byte) error {
 	return nil
 }
 
+// NullProto represents a Cloud Spanner PROTO that may be NULL.
+type NullProto struct {
+	ProtoVal proto.Message // ProtoVal contains the value when it is non-NULL, and nil when NULL.
+	Valid    bool          // Valid is true if Proto is not NULL.
+}
+
+// IsNull implements NullableValue.IsNull for NullProto.
+func (n NullProto) IsNull() bool {
+	return !n.Valid
+}
+
+// String implements Stringer.String for NullProto.
+func (n NullProto) String() string {
+	if !n.Valid {
+		return nullString
+	}
+	return n.ProtoVal.String()
+}
+
+// MarshalJSON implements json.Marshaler.MarshalJSON for NullProto.
+func (n NullProto) MarshalJSON() ([]byte, error) {
```

Comment:
Could you please help me understand when and how MarshalJSON is used?

Reply:
@Shua1 
We implement the `MarshalJSON()` and `UnmarshalJSON()` methods of `Marshaler` and `Unmarshaler` interfaces that belongs to the `json` package in golang. This is to provide the functionality of encoding custom types `NullProtoMessage` and `NullProtoEnum` into a `JSON` string and decoding `JSON` string into `NullProtoMessage` and `NullProtoEnum`.

**Usage:**
`json.Marshal` method can be used to encode a `NullProtoMessage` and `NullProtoEnum` into a JSON string.
`json.Unmarshal` method can be used to decode a JSON string into `NullProtoMessage` and `NullProtoEnum`.

As `NullProtoMessage` and `NullProtoEnum` are custom types we have to implement the above interfaces for the encode and decode to work as expected.

Reply:
Why do we need to encode and decode json?